rust/storage/src/caching.rs

```diff
@@ -0,0 +1,327 @@
+pub struct DedupeKey {
+    path: Path,
+}
+
+impl Key for DedupeKey {}
+
+#[derive(Debug, Default)]
+pub struct DedupeValue {
+    one_in_flight: tokio::sync::Mutex<()>,
+}
+
+impl Value for DedupeValue {
+    fn finished(&self) -> bool {
+        true
+    }
+}
+
+impl From<DedupeKey> for DedupeValue {
+    fn from(_: DedupeKey) -> Self {
+        Self::default()
+    }
+}
+
+#[derive(Clone)]
+pub struct CachingObjectStore {
+    cache: Arc<dyn ObjectStore>,
+    backing: Arc<dyn ObjectStore>,
+    dedupe: Arc<StateHashTable<DedupeKey, DedupeValue>>,
+}
+
+impl CachingObjectStore {
+    pub fn new<C: ObjectStore, B: ObjectStore>(cache: C, backing: B) -> Self {
+        Self {
+            cache: Arc::new(cache),
+            backing: Arc::new(backing),
+            dedupe: Arc::new(StateHashTable::new()),
+        }
+    }
+
+    async fn warm_cache(&self, location: &Path) -> Result<GetResult> {
+        let get = self
+            .backing
+            .get_opts(location, GetOptions::default())
+            .await?;
+        let meta = get.meta.clone();
+        let range = get.range.clone();
+        let attributes = get.attributes.clone();
+        let bytes = get.bytes().await?;
+        if let Err(err) = self
+            .cache
+            .put_opts(
+                location,
+                PutPayload::from(bytes.clone()),
+                PutOptions::default(),
+            )
+            .await
+        {
+            tracing::error!("failed to proactively warm the cache: {}", err);
+        }
+        let payload: GetResultPayload =
+            GetResultPayload::Stream(Box::pin(futures::stream::once(async move { Ok(bytes) })));
+        Ok(GetResult {
+            meta,
+            range,
+            attributes,
+            payload,
+        })
+    }
+}
+
+impl Debug for CachingObjectStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "CachingObjectStore")
+    }
+}
+
+impl Display for CachingObjectStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "CachingObjectStore")
+    }
+}
+
+#[async_trait::async_trait]
+impl ObjectStore for CachingObjectStore {
+    async fn put_opts(
+        &self,
+        location: &Path,
+        payload: PutPayload,
+        opts: PutOptions,
+    ) -> Result<PutResult> {
+        let res = self
+            .backing
+            .put_opts(location, payload.clone(), opts.clone())
+            .await?;
+        if let Err(err) = self
+            .cache
+            .put_opts(location, payload.clone(), opts.clone())
+            .await
+        {
+            tracing::error!("failed to proactively warm the cache: {}", err);
+        }
+        Ok(res)
+    }
+
+    async fn put_multipart_opts(
+        &self,
+        location: &Path,
+        opts: PutMultipartOpts,
+    ) -> Result<Box<dyn MultipartUpload>> {
+        self.backing.put_multipart_opts(location, opts).await
```

Comment:
In this case, is the client responsible for putting the data in the cache?

Reply:
The file will be put into cache at first load.  Unfortunately doing multi-part upload through to the cache is complicated and I took the easy path.  Happy to revisit if we see it's needed.  Alternatively we get to small blocks and use the non-multi-part.

Reply:
sounds good, request to just add a comment in the code about this divergence from non-multi-part